examples/bri-3/src/bri/communication/agents/bpiMessages.agent.ts

```diff
@@ -1,31 +1,101 @@
-import { Injectable, NotFoundException } from '@nestjs/common';
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
+import { LoggingService } from '../../../shared/logging/logging.service';
+import { BpiSubjectStorageAgent } from '../../identity/bpiSubjects/agents/bpiSubjectsStorage.agent';
+import { BpiSubject } from '../../identity/bpiSubjects/models/bpiSubject';
+import {
+  BPI_MESSAGE_ALREADY_EXISTS,
+  NOT_FOUND_ERR_MESSAGE,
+} from '../api/err.messages';
 import { BpiMessage } from '../models/bpiMessage';
 import { BpiMessageType } from '../models/bpiMessageType.enum';
-import { NOT_FOUND_ERR_MESSAGE } from '../api/err.messages';
 import { BpiMessageStorageAgent } from './bpiMessagesStorage.agent';
-import { BpiSubject } from '../../identity/bpiSubjects/models/bpiSubject';
-import { BpiSubjectStorageAgent } from '../../identity/bpiSubjects/agents/bpiSubjectsStorage.agent';
 
 @Injectable()
 export class BpiMessageAgent {
   constructor(
     private bpiMessageStorageAgent: BpiMessageStorageAgent,
     private bpiSubjectStorageAgent: BpiSubjectStorageAgent,
+    private readonly logger: LoggingService,
   ) {}
 
-  public async getFromAndToSubjectsAndThrowIfNotExist(
+  public async validateNewBpiMessageAgainstExistingBpiEntitiesWithThrow(
+    messageId: string,
     fromBpiSubjectId: string,
     toBpiSubjectId: string,
   ): Promise<[BpiSubject, BpiSubject]> {
+    const existingBpiMessage =
+      await this.bpiMessageStorageAgent.getBpiMessageById(messageId);
+
+    if (existingBpiMessage) {
+      throw new BadRequestException(BPI_MESSAGE_ALREADY_EXISTS);
+    }
+
     const fromBpiSubject = await this.bpiSubjectStorageAgent.getBpiSubjectById(
       fromBpiSubjectId,
     );
+
+    if (!fromBpiSubject) {
```

Comment:
shouldnt bpi subject storage agent throw this already? not strong opinion but we could eliminate this boilerplate everywhere if agent just throws not found

Reply:
This NotFoundException is a '@nestjs/common' thing that basically transalates to HTTP 404. When the command is invoked frrom something which is not HTTP, such as a NATS message, we need a different way to handle a stiutation where the resource is not found. This is why storage agents should be agnostic to the caller and just return null in case the resource is not there.